README.md
# ClashM
 Clash (Meta) in Magisk with Services.
 
 Official Telegram Channel [@ClashM_Magisk](https://t.me/ClashM_Magisk).

---
## Installation
  - Download installer zip file from [releases](https://github.com/adiprasetya/ClashM/releases).
  - Install it via Magisk Manager.

---
## Uninstallation
  - Uninstall modules via Magisk Manager.
  - Data directory automatically removed.

---
## Configuration
  - Data directory stored on `/data/adb/ClashM`
  - `base.yaml` base configuration.
  - `config.yaml` proxies configuration.
  - ClashM configuration at `/data/adb/modules/ClashM/scripts/configuration`
  - Run directory stored on `/data/adb/modules/ClashM/run`

---
## Credits
  - [@topjohnwu](https://github.com/topjohnwu) / Magisk
  - [@Dreamacro](https://github.com/Dreamacro) / clash
  - [@MetaCubeX](https://github.com/MetaCubeX) / Clash.Meta
  - [@haishanh](https://github.com/haishanh) / yacd
  - [@kalasutra](https://github.com/kalasutra) / Clash for Magisk
  - [@Asterisk4Magisk](https://github.com/Asterisk4Magisk) / Xray4Magisk
  - [@e58695](https://t.me/e58695)
  


changelog.md
---
## Changelogs
  - v1.2.2
    - Fix typo on customize.sh

  - v1.2.1
    - Add conditions on customize.sh.
    - Fix typo on start.sh.

  - v1.2.0
    - Removed `MERGE`.
    - New algorithm.
    - Add binary for arm32.
    - Update core.

  - v1.1.0
    - Allow hotspot connection toggle in configuration file (default true).
    - Rename clashm.sh to main.sh.
    - Added updateJson support.
  
  - v1.0.1
    - Change dashboard to [YACD MetaCubeX](https://github.com/MetaCubeX/yacd).
    - Update meta core.

  - v1.0.0
    - Removed TPROXY iptables support.
    - Removed useless/bloated features.
    - Rename `clashm.config` to `configuration`.
    - Cleaning up code.
    - Removed `geo-database` on zip installer.

  - v0.2.3
    - Add force option to skip port & tun device verifier.

  - v0.2.2
    - Clean up.
    - Fix get TUN interface on forward device.
  
  - v0.2.1
    - Using service.d.
    - Fix port handler.
    - Using TUN mode by default.

  - v0.2.0
    - Add notification.
    - DNS & TPROXY port handler.
    - Better tun device verification.
    - Update base, example configs.
    - Update yacd.

  - v0.1.1
    - Fix core.log output.
  
  - v0.1.0
    - Moved data dir to /data/adb.
    - Separate clashm.config in scripts dir.
    - Updated database.
    - Update meta core.
  
  - v0.0.1
    - Clean up code.
    - Update core.

  - v0.0.0
    - Initial release.



customize.sh
if [[ "$BOOTMODE" != true ]]; then
  abort "Install in Magisk Manager!"
fi

if [[ "$ARCH" != "arm"* ]]; then
  abort "- Only for arm devices!"
fi

if [[ "$IS64BIT" != "true" ]]; then
  ui_print "- Your devices isn't 64bit. PROCESS-NAME rules won't work"
fi

# variables
BIN="$MODPATH/bin"
DATA="/data/adb/$MODID"
SERVICE="/data/adb/service.d/$MODID.sh"
DATE="$(date "+%F")"

# rename binary
mv -f "$BIN/$ARCH" "$BIN/meta"
rm -f "$BIN/arm"*

ui_print "- Setup environment"
if [[ -f "$DATA/config.yaml" ]]; then
  cp -f "$DATA/config.yaml" "$MODPATH/data"
else 
  cp -f "$MODPATH/data/.example.yaml" "$MODPATH/data/config.yaml"
fi

if [[ -d "$DATA" ]]; then
  ui_print "- Old data directory exists,"
  ui_print "  Moved to $DATA.$DATE"
  if [[ -d "$DATA.$DATE" ]]; then
    ui_print "- Old $MODID.$DATE will replaced"
    rm -rf "$DATA.$DATE"
  fi
  mv -f "$DATA" "$DATA.$DATE"
fi

ui_print "- Data directory: $DATA"
mv -f "$MODPATH/data" "$DATA"
mkdir -p "$MODPATH/run"

ui_print "- Replacing variable"
sed -i "s|MODID|$MODID|" \
"$MODPATH/scripts/configuration" \
"$MODPATH/uninstall.sh" \
"$MODPATH/service.sh"

ui_print "- Installing service script"
mv -f "$MODPATH/service.sh" "$SERVICE"

ui_print "- Set environment permission"
set_perm_recursive "$MODPATH" 0 0 0755 0644
set_perm_recursive "$DATA" 0 0 0644 0644
set_perm_recursive "$MODPATH/scripts" 0 3005 0755 0755
set_perm_recursive "$MODPATH/bin" 0 3005 0755 0755
set_perm_recursive "$SERVICE" 0 0 0755 0755


scripts/start.sh
#!/system/bin/sh

DIR="${0%/*}"
. "$DIR/configuration"

wait_until_login() {
  # in case of /data encryption is disabled
  while [[ "$(getprop sys.boot_completed)" != "1" ]]; do
    sleep 1
  done

  # we doesn't have the permission to rw "/sdcard" before the user unlocks the screen
  local TEST_FILE="/sdcard/Android/.ClashM"
  while [[ ! -f "$TEST_FILE" ]]; do
    echo "true" > "$TEST_FILE"
    sleep 1
  done
  rm -f "$TEST_FILE"
}

start_service() {
  echo "true" > "$RUN_LOG"
  chmod 644 "$RUN_LOG"
  "$SCRIPTS/main.sh" start &> "$RUN_LOG"
}


main() {
  rm -f "$PID_FILE"
  chmod -R 0755 "$SCRIPTS"

  if [[ ! -f "$MODDIR/manual" ]]; then
    if [[ ! -f "$MODDIR/disable" ]]; then
      start_service
    fi
    inotifyd "$SCRIPTS/inotify.sh" "$MODDIR" &> /dev/null &
  fi
}


wait_until_login
main

scripts/clashm.sh
#!/system/bin/sh

DIR="${0%/*}"
. "$DIR/configuration"

TERMUX="/data/data/com.termux/files/usr/bin"
if [[ -d "$TERMUX" ]]; then
  PATH="$PATH:$TERMUX"
fi

# extracted from ( https://t.me/e58695/54 )
start_service() {
  if [[ ! -f "${BIN}" ]]; then
    echo "err: core is missing."
    exit 1
  fi

  if [[ ! -f ${CONFIG} ]]; then
    echo "err: configuration file does not exist!"
    exit 1
  fi

  chmod 0755 ${BIN}
  ${BIN} -t -d ${DATA} > ${DATA}/error.log
  if [[ "$?" != "0" ]]; then
    echo "err: configuration check failed!"
    exit 1
  fi

  echo "Date: $(date +%F)" > ${CORE_LOG_FILE}
  echo "Time: $(date +%R)" >> ${CORE_LOG_FILE}
  echo >> ${CORE_LOG_FILE}
  ulimit -SHn 1000000
  nohup ${BUSYBOX} setuidgid 0:3005 \
  "${BIN}" -d "${DATA}" -f "${CONFIG}"  &>> ${CORE_LOG_FILE} &
  echo -n $! > ${PID_FILE}
  echo "info: core started."
  rm -f ${DATA}/error.log
}

stop_service() {
  kill -15 `cat ${PID_FILE}` &> /dev/null
  rm -f ${PID_FILE}
  echo "info: core stopped."
}


tun_setup() {
  mkdir -p /dev/net
  if [[ ! -L /dev/net/tun ]]; then 
    ln -sf /dev/tun /dev/net/tun
  fi
}

merger() {
  [[ "$MERGE" == "true" ]] || return
  echo "info: force merge."
  cp -f "$BASE" "$CONFIG"
  echo >> "$CONFIG"
  cat "$PROXIES" >> "$CONFIG"
}

forward_device() {
  device="$(awk '/device/ {print $2}' "$CONFIG")"
  interface=(utun Meta $device)
  for i in ${interface[@]}; do
    iptables -I FORWARD -o "$i" -j ACCEPT
    iptables -I FORWARD -i "$i" -j ACCEPT
    ip6tables -I FORWARD -o "$i" -j ACCEPT
    ip6tables -I FORWARD -i "$i" -j ACCEPT
  done
}


start() {
  local pid=`cat ${PID_FILE} 2> /dev/null`
  if (cat /proc/${pid}/cmdline | grep -q ${BIN}); then
    echo "info: core has been started."
    exit 1
  fi

  merger
  tun_setup
  start_service
  forward_device
}


case "$1" in
  start|s)
    start
   ;;
  stop|k)
    stop_service
   ;;
  *)
    echo "Usage: $0 {start|stop}"
   ;;
esac


scripts/main.sh
#!/system/bin/sh

DIR="${0%/*}"
. "$DIR/configuration"

# extracted from ( https://t.me/e58695/54 )
start_service() {
  if [[ ! -f "$BIN" ]]; then
    echo "error: core is missing."
    exit 1
  fi

  if [[ ! -f "$TEMPORARY" ]]; then
    echo "error: configuration file does not exist!"
    exit 1
  fi

  ERROR_LOG="$DATA/error.log"
  chmod -R 0755 "$BIN_DIR"
  "$BIN" -t -d "$DATA" -f "$TEMPORARY" > "$ERROR_LOG"
  if [[ "$?" != "0" ]]; then
    echo "error: configuration check failed!"
    echo "help: check $ERROR_LOG for clue, and re-check configuration file."
    exit 1
  fi

  date "+%A, %d %B %Y | %R %Z" > "$CORE_LOG_FILE"
  echo >> "$CORE_LOG_FILE"
  ulimit -SHn 1000000
  nohup "$BUSYBOX" setuidgid 0:3005 \
  "$BIN" -d "$DATA" -f "$TEMPORARY"  &>> "$CORE_LOG_FILE" &
  echo -n "$!" > "$PID_FILE"
  echo "info: core started."
  rm -f "$ERROR_LOG"
}

stop_service() {
  PID="$(cat "$PID_FILE")"
  kill -15 "$PID" &> /dev/null
  rm -f "$PID_FILE"
  rm -f "$TEMPORARY"
  echo "info: core stopped."
}


android_tun() {
  mkdir -p /dev/net
  if [[ ! -L /dev/net/tun ]]; then 
    ln -sf /dev/tun /dev/net/tun
  fi
}

merge_config() {
  if [[ ! -f "$BASE" ]]; then
    echo "info: config not merged."
    TEMPORARY="$CONFIG"
    return 0
  fi
  cp -f "$BASE" "$TEMPORARY"
  echo >> "$TEMPORARY"
  cat "$CONFIG" >> "$TEMPORARY"
}

hotspot() {
  device="$(awk '/device:/ {print $2}' "$TEMPORARY")"
  interface=(utun Meta $device)
  if [[ "$HOTSPOT" != "ACCEPT" ]]; then
    HOTSPOT="REJECT"
    echo "info: hotspot traffic rejected."
  fi
  for i in ${interface[@]}; do
    iptables -I FORWARD -o "$i" -j "$HOTSPOT"
    iptables -I FORWARD -i "$i" -j "$HOTSPOT"
    ip6tables -I FORWARD -o "$i" -j "$HOTSPOT"
    ip6tables -I FORWARD -i "$i" -j "$HOTSPOT"
  done
}


start() {
  local pid=`cat "$PID_FILE" 2> /dev/null`
  if (cat "/proc/$pid/cmdline" | grep -q "$BIN"); then
    echo "info: core has been started."
    exit 1
  fi

  android_tun
  merge_config
  start_service
  hotspot
}


case "$1" in
  start|s)
    start
   ;;
  stop|k)
    stop_service
   ;;
  *)
    echo "Usage: $0 {start|stop}"
   ;;
esac
